x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adapitiveallocations/AdaptiveAllocationsScaler.java

```diff
@@ -0,0 +1,140 @@
+        this.maxNumberOfAllocations = null;
+        this.dynamicsChanged = false;
+    }
+
+    void setMinMaxNumberOfAllocations(Integer minNumberOfAllocations, Integer maxNumberOfAllocations) {
+        this.minNumberOfAllocations = minNumberOfAllocations;
+        this.maxNumberOfAllocations = maxNumberOfAllocations;
+    }
+
+    void process(AdaptiveAllocationsScalerService.Stats stats, double timeIntervalSeconds, int numberOfAllocations) {
+        // The request rate (per second) is the request count divided by the time.
+        // Assuming a Poisson process for the requests, the variance in the request
+        // count equals the mean request count, and the variance in the request rate
+        // equals that variance divided by the time interval squared.
+        // The minimum request count is set to 1, because lower request counts can't
+        // be reliably measured.
+        // The estimated request rate should be used for the variance calculations,
+        // because the measured request rate gives biased estimates.
+        double requestRate = (double) stats.requestCount() / timeIntervalSeconds;
+        double requestRateEstimate = requestRateEstimator.hasValue() ? requestRateEstimator.estimate() : requestRate;
+        double requestRateVariance = Math.max(1.0, requestRateEstimate * timeIntervalSeconds) / Math.pow(timeIntervalSeconds, 2);
+        requestRateEstimator.add(requestRate, requestRateVariance, false);
+
+        if (stats.requestCount() > 0) {
+            // The inference time distribution is unknown. For simplicity, we assume
+            // a std.error equal to the mean, so that the variance equals the mean
+            // value squared. The measurement variance scales inversely proportional
+            // to the number of measurements.
+            // Again, the estimated inference time should be used for the variance
+            // calculations to prevent biased estimates.
+            double inferenceTime = stats.inferenceTime();
+            double inferenceTimeEstimate = inferenceTimeEstimator.hasValue() ? inferenceTimeEstimator.estimate() : inferenceTime;
+            double inferenceTimeVariance = Math.pow(inferenceTimeEstimate, 2) / stats.requestCount();
+            inferenceTimeEstimator.add(inferenceTime, inferenceTimeVariance, dynamicsChanged);
+        }
+
+        this.numberOfAllocations = numberOfAllocations;
+        dynamicsChanged = false;
+    }
+
+    Integer scale() {
+        if (requestRateEstimator.hasValue() == false) {
+            return null;
+        }
+
+        int oldNumberOfAllocations = numberOfAllocations;
+
+        double requestRateLower = Math.max(0.0, requestRateEstimator.lower());
+        double inferenceTimeLower = Math.max(0.0, inferenceTimeEstimator.hasValue() ? inferenceTimeEstimator.lower() : 1.0);
+        double loadLower = requestRateLower * inferenceTimeLower;
+        while (loadLower / numberOfAllocations > SCALE_UP_THRESHOLD) {
```

Comment:
I think (and iirc observed) that noisy initial measurements / very few data points can cause unnecessary scaling up, because of a load estimate with large error. Therefore I think this current behaviour is better.

The threshold of 90% should take care of preventing `"true load" / numberOfAllocations > 1`.

Reply:
Ok, probably the parameters work out in our favour and the estimate's standard deviation << 0.1. I can buy that without actually working through the details. This will be tied up with the amount of smoothing we use though. For example if this is reduced it will likely cause problems. Perhaps it might be worth adding a test `requestRateLower * inferenceTimeLower - requestRateEstimator.estimate() * inferenceTimeEstimator.estimate()` is typically less than `1 - SCALE_UP_THRESHOLD`. This would fail if anyone modified the code in such a way as to invalidate this assumption.

Reply:
OK, added one more test that I think tests this. Let me know what you think...

Reply:
I don't think you pushed this yet

Reply:
🤦